### PR TITLE
Initial Sparse Matrix Support

### PR DESCRIPTION
**Pull request by kuza55**

### CODE REVIEW

**Comment by kuza55** [comment]
So, this adds the minimal changes necessary to Keras so that a net can be trained with a scipy.sparse matrix

This probably also needs a sparse input layer, but I thought I would get some feedback on this before throwing too much time into it.

This is the code I used to test this:

```
import numpy as np
import keras
from keras.layers import Input, Dense
from keras.models import Model
import tensorflow as tf

from scipy.sparse import csr_matrix

#Setup sparse input
x_d = np.array([0, 7, 2, 3], dtype=np.int64)
x_r = np.array([0, 2, 2, 3], dtype=np.int64)
x_c = np.array([4, 3, 2, 3], dtype=np.int64)

y = np.array([0, 1, 2, 3], dtype=np.int64)

x_in = csr_matrix((x_d, (x_r, x_c)), shape=(4, 5))

#Setup Model
numeric_columns = 5

sparse = Input(batch_shape=(None, numeric_columns,), tensor=tf.sparse_placeholder(tf.float32))

predictions = Dense(1, activation='sigmoid')(sparse)

optimizer = keras.optimizers.Adam()
model = Model(input=[sparse], output=predictions)
model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

#Run it!
model.fit(x_in, y)
```

**Comment by fchollet** [comment]
Looks reasonable. What happens to your example with Theano? Or would Theano support have to wait for a cross-backend sparse input layer.

**Comment by kuza55** [comment]
So at the moment, if you remove the sparse_placeholder tensor and try and feed in a sparse matrix to the theano backend it crashes when trying to convert the sparse matrix to an array.

One thing that could be done to both backends is to have them auto-convert sparse matrixes to dense matrixes per-batch when they see dense placeholders. It might be useful for large data that people don't want to write generators for.

I know nothing about theano, but from the reading I just did it looks like it works roughly the same way: you create a sparse placeholder and then you can pass in a sci-py matrix into the theano function directly: http://deeplearning.net/software/theano/tutorial/sparse.html

I'll see if I can hack together a theano version.

**Comment by kuza55** [comment]
So that turned out to be relatively straight forward to get TF parity; one thing I noticed is that Theano has many more sparse operations, so there were 3 kinds of different dot products you could choose from. I chose one that produces a dense output and doesn't compute gradients for empty parts of the sparse matrix; you can see the descriptions if you scroll down here: http://deeplearning.net/software/theano/library/sparse/index.html

Theano also supports more sparse formats, though for batch training I expect columns > rows for cases where you actually want a sparse matrix, so I think that using a CSR representation by default in a sparse input layer is probably fine.

This is the test code I used that ran:

```
import numpy as np
import keras
from keras.layers import Input, Dense
from keras.models import Model
import theano.sparse as th_sp

from scipy.sparse import csr_matrix

#Setup sparse input
x_d = np.array([0, 7, 2, 3], dtype=np.int64)
x_r = np.array([0, 2, 2, 3], dtype=np.int64)
x_c = np.array([4, 3, 2, 3], dtype=np.int64)

y = np.array([0, 1, 2, 3], dtype=np.int64)

x_in = csr_matrix((x_d, (x_r, x_c)), shape=(4, 5))

#Setup Model
numeric_columns = 5

tensor=th_sp.csr_matrix(name='sparse_input', dtype='float32')

sparse = Input(batch_shape=(None, numeric_columns,), tensor=tensor)

predictions = Dense(1, activation='sigmoid')(sparse)

optimizer = keras.optimizers.Adam()
model = Model(input=[sparse], output=predictions)
model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

#Run it!
model.fit(x_in, y)
```

**Comment by kuza55** [comment]
Yep, looks like the nose-parameterized dep causes travis to fail. I could try to catch the import error and don't allow sparse input creation if the dependency failed to load, let me know if you think it's worth it.

**Comment by kuza55** [comment]
So I added the sparse Input support and also added sparse concat support for TF. I'm not super keen on adding Theano support myself since I don't have any need and I don't know Theano, but I'm not sure how you feel about having things that only support one backend.

I'm running into a few issues that make this harder:
a) SparseTensor doesn't have a get_shape() function
b) sparse_placeholder won't take normal shape arguments - https://github.com/tensorflow/tensorflow/issues/4226
b) Sci-Py sparse matrixes don't support len() - https://github.com/scipy/scipy/issues/6549

So far a) has been possible to work around since code has mostly just wanted the rank rather than the full shape; I wonder if I should make the code in TF_K.concatenate call into TF_K.ndim for this?

b) has been possible to resolve, but only to the extent that code only really needs the rank

c) I've been replacing calls to len(mat) with calls to mat.shape[0], but I've also filed a sci-py ticket where people seem responsive.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -113,6 +113,32 @@ def _to_tensor(x, dtype):
     return x
 
 
+def is_sparse(tensor):
+    return isinstance(tensor, tf.SparseTensor)
+
+
+def how_sparse(tensors):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Unnecessary. Use `all([K.is_sparse(x) for x in tensors])` or `any([K.is_sparse(x) for x in tensors])`. Far more readable than an obscure API call for which you would have to read the source to understand what it does.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -113,6 +113,32 @@ def _to_tensor(x, dtype):
...
+    for tensor in tensors:
+        if is_sparse(tensor):
+            any_sparse = True
+        else:
+            all_sparse = False
+
+    return (any_sparse, all_sparse and any_sparse)
+
+
+def to_dense(tensors):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Sure this should take a single tensor as input for API consistency. Using `[K.to_dense(x) for x in tensors]` is also not a big overhead for the batch version.

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -5,6 +5,11 @@
 from theano.tensor.nnet import conv3d2d
 from theano.printing import Print
 try:
+    import theano.sparse as T_sp
+    HAS_SPARSE = True
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: `T_sp` should be renamed to something readable (like `th_sparse_module`). No need for an additional global variable, just set `th_sparse_module` to None if not available.

File: keras/engine/topology.py [hunk 4]
```diff
@@ -947,7 +947,7 @@ class InputLayer(Layer):
     '''TODO: dosctring
     '''
     def __init__(self, input_shape=None, batch_input_shape=None,
-                 input_dtype=None, input_tensor=None, name=None):
+                 input_dtype=None, input_tensor=None, sparse=None, name=None):
```
Comment (fchollet) [inline keras/engine/topology.py]: Should `sparse` be a boolean?

**Comment by kuza55** [comment]
@fchollet I addressed your comments as best I could; though using all() wasn't possible since it gets shadowed by K.all, so I copied it to py_all, hopefully that's better than before, lmk if you have a preferred solution.

Do you have any thoughts for making this work for Sequential models? I don't really use them, so maybe it's not even important, but it looks like allowing sparse inputs there would require adding an argument to Model, otherwise it would require changing every class, which seems like no fun.

The main issue I see cropping up is the scipy __len__ weirdness, but hopefully that can get resolved in scipy and we can revert the changes I made to training.py to get it to work.

**Comment by kuza55** [comment]
So it looks like I wrote some flakey tests.

In particular it seems that sometimes sometimes sparse multiplication can be slightly off from the dense results:

```
>           assert np.all(k_s == k_d)
E           assert <function all at 0x7fe76ed31140>(array([[ 0.  ...dtype=float32) == array([[ 0.   ...dtype=float32)
E            +  where <function all at 0x7fe76ed31140> = np.all
E             Full diff:
E             array([[ 0.        ,  0.        ,  0.        ,  0.        ],
E             [ 0.        ,  0.        ,  0.        ,  0.        ],
E             -        [ 7.66850376,  6.24425411,  4.40544653,  4.35426092],
E             ?                  ^^
E             +        [ 7.66850328,  6.24425411,  4.40544653,  4.35426092],
E             ?                  ^^
E             [ 2.9856596 ,  2.03145814,  1.55561113,  1.47568619]], dtype=float32))
```

Do you have any advice on how to fix this? Should I subtract the two and check that they're all within some margin of error? Should I remove the equality test and just check that the result was the right shape?

**Comment by fchollet** [comment]
Seed your RNG to avoid indeterminism. Use `assert_allclose` to compare Numpy arrays, with an appropriate threshold (e.g. 1e-5). http://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.testing.assert_allclose.html

**Comment by kuza55** [comment]
@fchollet The tests are fixed now. It should probably be ready to merge now.

The main change that I think is missing is to fix scipy's sparse matrix __len__ support so that we don't randomly blow up when using sparse matrixes in places I haven't patched up, but I don't think that needs to be a blocker to this getting merged.

File: keras/backend/theano_backend.py [hunk 5]
```diff
@@ -30,17 +35,28 @@ def set_learning_phase(value):
...
+    else:
+        return tensor
+
 def variable(value, dtype=_FLOATX, name=None):
     '''Instantiate a tensor variable.
     '''
-    value = np.asarray(value, dtype=dtype)
-    return theano.shared(value=value, name=name, strict=False)
+    if hasattr(value, 'tocoo'):
+        return th_sparse_module.as_sparse_variable(value)
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: This would trigger a hard-to-read exception if the module is not available. Better to throw a proper exception with an appropriate message.

**Comment by fchollet** [comment]
LGTM, thanks
